### ISSUE TITLE
IDE0044 incorrectly identifies generic types as immutable

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#37981]
**Version Used**: 
3.3.0-beta3-19409-05+ed92d532473db83c1db47b313ee1c1bd7520aa08

**Steps to Reproduce**:
Copy this into a new project:
```C#
using System;
using System.Collections;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        var e = Wrap(new List<int> { 1, 2, 3, 4, 5 }.GetEnumerator());
        while (e.MoveNext()) Console.WriteLine(e.Current);
    }

    private static EnumeratorWrapper<T> Wrap<T>(T enumerator)
        where T : IEnumerator<int> =>
        new EnumeratorWrapper<T>(enumerator);
}

internal sealed class EnumeratorWrapper<T> :
    IEnumerator<int>
    where T : IEnumerator<int>
{
    private T _enumerator;

    public EnumeratorWrapper(T enumerator) => _enumerator = enumerator;

    public int Current => _enumerator.Current;
    object? IEnumerator.Current => _enumerator.Current;
    public void Dispose() => _enumerator.Dispose();
    public bool MoveNext() => _enumerator.MoveNext();
    public void Reset() => _enumerator.Reset();
}
```
Run it.  As expected, it prints out:
```
1
2
3
4
5
```
Then apply IDE0044, which triggers for the `_enumerator` field, and run it.  The program will never end, repeatedly printing out an infinite number of `0`s.

**Expected Behavior**:
IDE0044 should not trigger for the `_enumerator` field.  The rule has logic to identify fields known to be of types that are either reference types or immutable structs, and it needs to be conservative in its estimation of whether a struct is immutable, always erring on the side of mutable if it can't be sure, since automatically making a field of a mutable struct type readonly can lead to erroneous behavior, as the repro demonstrates.  The rule should consider generic parameters to be mutable, unless they're constrained to be reference types.

**Actual Behavior**:
IDE0044 triggers for `_enumerator`, marking the field `readonly`, and leading to buggy program behavior.

**Comment by sharwell** [comment]
> The rule should consider generic parameters to be mutable, unless they're constrained to be reference types.

It would probably also continue to treat _unconstrained_ generic types as immutable. Technically `ToString` and `GetHashCode` could be overridden to include mutating code on a value type, but this would be an extremely rare edge case that would substantially impact the usability of the diagnostic.

**Comment by stephentoub** [comment]
> It would probably also continue to treat unconstrained generic types as immutable

Right, that's why I said "unless they're constrained to be reference types".

**Comment by stephentoub** [comment]
> It would probably also continue to treat unconstrained generic types as immutable

You mean because then ToString and GetHashCode are the only operations that could be performed on the instance?

**Comment by stephentoub** [comment]
> but this would be an extremely rare edge case

It's not that rare... a fair number of ToString and GetHashCode implementations cache the result for subsequent use.  Marking such fields as readonly would inhibit that caching.

**Comment by stephentoub** [comment]
FWIW, the benefits of marking a field `readonly` are small enough and the potential downside large enough that I think automatic refactorings should be as conservative as possible, and I'd suggest eliminating from consideration all generic types not constrained to `class`.

Maybe you should consider adding the ability for rules to specify certainty of some kind, if that capability doesn't already exist.  IDE0044 could then automatically make readonly those it was "sure" about, and could suggest for consideration those it was less sure about.

**Comment by stephentoub** [comment]
Also, just FYI, I found this because in trying to apply IDE0044 to corefx, this introduced several bugs that were then thankfully caught by our tests.

**Comment by CyrusNajmabadi** [comment]
>  the potential downside large enough that I think automatic refactorings should be as conservative as possible, and I'd suggest eliminating from consideration all generic types not constrained to class.

I personally disagree.  We've taken that approach in the past and making things as conservative as possible not only made things often show up never (because actually being able to prove something is totally safe is exceedingly rare), but it also impacted resharper compete.  People have shown they're willing to accept a level of balance between things working 99% of the time, but possibly breaking in the 1% case, versus only being offered in the 1% case where it can be totally proven to be safe.

This is a delicate balance, and we have tools at our disposable to improve the experience here.  For example, one option is to offer here, but give warning that there's an issue.  Another is to have fix all be slightly more conservative and require explicit consent on these cases.

> FWIW, the benefits of marking a field readonly are small enough and

I would personally disagree.  Making fields readonly is one of the top things i find code can do to improve its health.  Being able to reason and work with types that are closer and close to fully immutable types is a huge book to understanding and code health.  I think we need to strike a good balance.  But we should be attempting to help people move to this, while being cognizant of the risks in some cases.

**Comment by CyrusNajmabadi** [comment]
> Also, just FYI, I found this because in trying to apply IDE0044 to corefx, this introduced several bugs that were then thankfully caught by our tests.

Note: from 2007 onwards, this was the explicit design choice picked by the IDE.  We moved away from a conservative 'only safe' approach due to such a large amount of feedback about the problems inherent therein.  There's upwards of a decade of experience with this that has shown that this is *definitely* a problem and a concern but *not* that it's such a concern or issue to warrant swinging back in hte other direction.

**Comment by stephentoub** [comment]
> People have shown they're willing to accept a level of balance between things working 99% of the time, but possibly breaking in the 1% case, versus only being offered in the 1% case where it can be totally proven to be safe.

It may just be then that for a large codebase like corefx, we shouldn't be using these rules.  The number of times the rules have introduced breaks is high enough that I've had to painstakingly review each and every change the rules make.  And at this point the return on investment is diminishing to the point where, since I can't trust the refactorings to be 100% accurate, I'll just stop using them.

> Making fields readonly is one of the top things i find code can do to improve its health.

Don't get me wrong, I really like making things readonly.  But making things readonly can have silent consequences, either making things functionally incorrect or introducing copies that impact perf.  This is one place I think we're doing ourselves and everyone else a disservice by not being as accurate as we can.

**Comment by CyrusNajmabadi** [comment]
> It may just be then that for a large codebase like corefx, we shouldn't be using these rules.

Yes.  The addition of them as rules does throw a wrinkle into things.  I'm not sure the best way to balance all the needs here.  It's possible that an appropriate thing to do here is *not* support fix-all, or make fix-all only operate on safer things so that users have to think through the implications here.  

We can and do support the concept of warnings with fixes (which at the very minimum should be here).  Perhaps that, combined with a policy of "do not 'fix all' fixes with warnings" is a good starting point to find a better balance here?

**Comment by CyrusNajmabadi** [comment]
> Don't get me wrong

Definitely understood.  And your feedback is fair, critical and appropriate.  I'm really glad you opened this.  I'm just adding some additional context to help feed into the discussion.  To be clear, I do *not* think we're in a good state here, and i'm really bummed you were negatively impacted here.  I'm just trying to add information so we can brainstorm potential middle-round fixes before going with the much more nuclear approach (if that makes sense).

**Comment by stephentoub** [comment]
> Perhaps that, combined with a policy of "do not 'fix all' fixes with warnings" is a good starting point to find a better balance here?

I'd be fine with that; this is akin to what I wrote above: "Maybe you should consider adding the ability for rules to specify certainty of some kind, if that capability doesn't already exist. IDE0044 could then automatically make readonly those it was "sure" about, and could suggest for consideration those it was less sure about, letting the developer decide."

I would like the option of saying "Fix everything that can be fixed with 100% fidelity", and then as a nice to have ideally "Show me everything else you think can be fixed but are less sure about".  IIRC, FxCop had a "certainty" mechanism; it'd be nice if there was such a thing here, and then in addition to being able to set an action level for a rule, I could instead do so for certainty levels, e.g. "Suggestion" at 100%, "Refactoring" at 80%, and everything below that "None".

**Comment by CyrusNajmabadi** [comment]
Very interesting ideas.  It would also give us a way to do a 'bar' check.  i.e. every time we run into something where we have to say "oh man... this is super super useful, but not 100% correct if we have to consider every single corner case" we can at least use this mechanism to identify and classify, leaving it in the hands of the user to decide what policy makes the most sense to them.

Here's a good example of the types of things this would stop from working, but may be ok from a user perspective:

A huge chunk of our fixes look for unnecessary redundancy in code.  i.e. find `return x == null ? null : x.y.z.w` and return `return x?.y.z.w;`.  However, when i ran this in roslyn, technically 99% of all cases we fixed up were technically "unsafe", but in practice were "safe".  That's because we couldn't prove that 'x' as an expr didn't have side-effects (and thus going from two invokes of it to one could break things).  But, from analyzing codebases (including corefx) we couldn't find *any* cases where people might have side effects that this would be a problem for.

So *at the time* what was the right thing to do there? Be conservative, block 99% of the use cases (even though everything we examined was actually safe in all those cases), and make the feature far less useful for everyone?  Or say that the expectation was that if enough of these checks pass, that we have enough confidence to recommend to the user.  

--

Now, if we tweak things to leave this choice in the user's hands, that above example can now have the following logic: set the bit saying if we're 100% certain.  In the case where 'x' may cause side-effects, lower the certainty.    You could now have this analyzer telling you about these cases.  But you'd have to examine and decide on a case by bcase basis with your decision to not fix things at less than 100% confidence.

--

I def think there's a design space to explore here.  @sharwell @kendrahavens  can this topic be taken to an ide design meeting?
